keps/sig-storage/20190117-SkipVolumeOwnershipChange.md

```diff
@@ -0,0 +1,147 @@
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+## Summary
+
+Currently before a volume is bind-mounted inside a container the permissions on
+that volume are changed recursively to the provided fsGroup value.  This change
+in ownership can take an excessively long time to complete, especially for very
+large volumes (>=1TB) as well as a few other reasons detailed in [Motivation].
+To solve this issue we will add a new field called `PermissionChangePolicy` and
+allow the user to specify whether they want the ownership change to occur.
+
+## Motivation
+
+When a volume is mounted on the node, we recursively change permissions of volume
+before bind mounting the volume inside container. The reason of doing this is to ensure
+that volumes are readable/writable by provided fsGroup.
+
+But this presents following problems:
+ - An application(many popular databases) which is sensitive to permission bits changing
+   underneath may refuse to start whenever volume being used inside pod gets mounted on
+   different node.
+ - If volume has a large number of files, performing recursive `chown` and `chmod`
+   could be slow and could cause timeout while starting the pod.
+
+### Goals
+
+ - Allow volume ownership and permission to be skipped during mount
+
+### Non-Goals
+
+ - In some cases if user brings in a large enough volume from outside, the first time ownership and permission change still could take lot of time.
+ - On SELinux enabled distributions we will still do recursive chcon whenever applicable and handling that is outside the scope.
+
+## Proposal
+
+We propose that an user can optionally opt-in to skip recursive ownership(and permission) change on the volume if volume already has right permissions.
+
+### Implementation Details/Notes/Constraints [optional]
+
+When creating a pod, we propose that `PersistentVolumeClaimVolumeSource` field expanded to include a new field called `PermissionChangePolicy` which can have following possible values:
+
+ - `NoChange` --> Don't change permissions and ownership.
+ - `Always` --> Always change the permissions and ownership to match fsGroup. This is the current behavior and it will be the default one when this proposal is implemented.
+ - `OnDemand` --> Only change permissions when fsGroup of pod and PVC don't match.
+
+```go
+type PersistentVolumeClaimVolumeSource struct {
+    // ClaimName is the name of a PersistentVolumeClaim in the same namespace as the pod using this volume
+    ClaimName string
+    // Optional: Defaults to false (read/write).  ReadOnly here
+    // will force the ReadOnly setting in VolumeMounts
+    // +optional
+    ReadOnly bool
+    // PermissionChangePolicy ← new field
+    PermissionChangePolicy string
```

Comment:
What happens if multiple pods that share the same PVC specify conflicting policies?  fsGroup has this same issue because it's specified at the Pod level, and not the PVC level.  Should we just say "undefined", or try to actually enforce consistency?

Reply:
Yeah so mount operations are serialized on the node if multiple pods use same PVC and in which case last pod to run wins. We can document this fact but I would imagine this should be no different than how existing `fsGroup` is handled on mount.